server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+      }
+    } catch (Exception e) {
+      LOG.error("Exception happened when getShuffleData in buffer", e);
+    }
+    return new ShuffleDataResult();
+  }
+
+  // here is the rule to read data in memory:
+  // 1. read from inFlushBlockMap order by eventId asc, then from blocks
+  // 2. if can't find lastBlockId, means related data may be flushed to storage, repeat step 1
+  private ActualReadInfo updateBufferSegments(
+      long lastBlockId,
+      long readBufferSize,
+      List<BufferSegment> bufferSegments) throws Exception {
+    long fromEventId = INVALID_EVENT_ID;
+    long currentBlockId = lastBlockId;
+    List<Long> sortedEventId = sortFlushingEventId();
+    int offset = 0;
+    int eventIndex = 0;
+    long updatedLastBlockId = lastBlockId;
+    boolean hasBlockId = false;
+    // read from inFlushBlockMap first
+    if (inFlushBlockMap.size() > 0) {
+      for (Long eventId : sortedEventId) {
+        ListReadInfo listReadInfo = processCachedBlocks(offset, inFlushBlockMap.get(eventId),
+            currentBlockId, readBufferSize, bufferSegments);
+        offset = listReadInfo.offset;
+        // try set which eventId is the start point
+        if (fromEventId == INVALID_EVENT_ID) {
+          if (offset > 0) {
+            // offset > 0, start from current event
+            fromEventId = eventId;
+            // read from 1st block with following inFlushBlock/blocks
+            currentBlockId = Constants.INVALID_BLOCK_ID;
+          } else if (listReadInfo.hasBlockId) {
+            // offset = 0, but find the blockId, start from next event if exist
+            if (eventIndex < sortedEventId.size() - 1) {
+              fromEventId = sortedEventId.get(eventIndex + 1);
+            }
+            hasBlockId = true;
+            // read from 1st block with following inFlushBlock/blocks
+            currentBlockId = Constants.INVALID_BLOCK_ID;
+            updatedLastBlockId = Constants.INVALID_BLOCK_ID;
+          }
+        }
+        if (offset >= readBufferSize) {
+          break;
+        }
+        eventIndex++;
+      }
+    }
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      ListReadInfo listReadInfo = processCachedBlocks(
+          offset, blocks, currentBlockId, readBufferSize, bufferSegments);
+      offset = listReadInfo.offset;
+      hasBlockId = listReadInfo.hasBlockId;
+    }
+    if (lastBlockId != Constants.INVALID_BLOCK_ID && offset == 0 && !hasBlockId) {
+      // can't find lastBlockId, it should be flushed
+      // try read with blockId = Constants.INVALID_BLOCK_ID
+      return updateBufferSegments(Constants.INVALID_BLOCK_ID, readBufferSize, bufferSegments);
+    } else {
+      return new ActualReadInfo(offset, fromEventId, updatedLastBlockId);
+    }
+  }
+
+  private void updateShuffleData(
+      long lastBlockId,
+      long readBufferSize,
+      byte[] data,
+      long fromEventId) throws Exception {
+    List<Long> sortedEventId = sortFlushingEventId();
```

Comment:
We sort events multi times. If event is flushed. The views will be different.

Reply:
It's ok, read won't depend on the view status, just get data according to view & blockId. The backward is duplicated data may be read.